### PR TITLE
Properly validate offset in GridWrap.ScrollToOffset

### LINKED ISSUE #4655: GridWrap ScrollToOffset API doesn't properly validate offset when viewport is larger than the full content
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [X] This issue only relates to a single bug. I will open new issues for any other problems.

### Describe the bug

If a gridwrap is showing items in a viewport larger than needed to show everything in one screen, ScrollToOffset doesn't validate the offset correctly and you can scroll "down", causing items to disappear from the viewport, and "up" in a way that can result in a crash

### How to reproduce

It shows up in Supersonic on main (development) branch now that I've added keyboard scrolling. May add example code later...

### Screenshots

_No response_

### Example code

might add later

### Fyne version

2.4.4

### Go compiler version

go1.21.1

### Operating system and version

macOS Ventura M2

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by dweymouth**
### Description:
Ensures that ScrollToOffset cannot be used to scroll beyond the content size, even if the viewport (ie size of the widget itself) is larger than the content size. Also fixes a missed Refresh call in ScrollToOffset.

Fixes #4655 

### Checklist:

- [x] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.

### CODE REVIEW

File: widget/gridwrap.go [hunk 1]
```diff
@@ -317,6 +326,17 @@ func (l *GridWrap) UnselectAll() {
...
 
+func (l *GridWrap) contentMinSize() fyne.Size {
+	padding := theme.Padding()
+	if lenF := l.Length; lenF != nil {
+		cols := l.ColumnCount()
+		rows := float32(math.Ceil(float64(lenF()) / float64(cols)))
+		return fyne.NewSize(l.itemMin.Width,
+			(l.itemMin.Height+padding)*rows-padding)
+	}
+	return fyne.NewSize(0, 0)
```
Comment (Jacalz) [inline widget/gridwrap.go:337]: I realise that this just was copied from the other method more or less but lets get the logic ordered up to our usual standard with edge-cases indented and main path inline:
```suggestion
	if l.Length == nil {
		return fyne.NewSize(0, 0)
	}
	
	cols := l.ColumnCount()
	rows := float32(math.Ceil(float64(l.Length()) / float64(cols)))
	return fyne.NewSize(l.itemMin.Width, (l.itemMin.Height+padding)*rows-padding)
```

**Review by Jacalz — CHANGES_REQUESTED** [review]
Sorry. I realised that the code wasn't conforming to our style guides. I know it is more or less fixing up old code but lets get it solved now that you are touching it :)

File: widget/gridwrap.go (hunk 1 above)
Comment (dweymouth) [reply widget/gridwrap.go:337]: Interestingly, the old way was more thread-safe than the new way (user could unset the Length callback in between checking it for nil), but given that we are moving to the single-threaded model it's not an issue.

Comment (Jacalz) [reply widget/gridwrap.go]: Without actually using any locks they are both half broken but yeah the old one was better in that regard. You can always save the callback to a variable if you like.

Comment (dweymouth) [reply widget/gridwrap.go]: Nah, it won't be an issue at all with the new threading model :) (b/c you'll be disallowed from changing it while the logic here is running)

Comment (Jacalz) [reply widget/gridwrap.go]: But you have labelled this as going into v2.4.5?

Comment (dweymouth) [reply widget/gridwrap.go]: As you mentioned it was half-broken anyway, and unsetting the Length callback after having set it before for List is basically a developer error anyway (I can't think of a legitimate use case). I wouldn't worry about it, but if you want I can make the change

**Review by Jacalz — APPROVED** [review]

**Comment by dweymouth** [comment]
@Jacalz needs re-approval since there was a trailing whitespace picked up that failed static analysis :)

**Review by Jacalz — APPROVED** [review]

**Pull request merged**